pkg/connector/user.go

```diff
@@ -0,0 +1,252 @@
+//   - annotations.Annotations: Additional metadata
+//   - error: Any error that occurred during the operation
+func (o *userBuilder) Grants(ctx context.Context, resource *v2.Resource, pToken *pagination.Token) ([]*v2.Grant, string, annotations.Annotations, error) {
+	var grants []*v2.Grant
+	annos := annotations.Annotations{}
+
+	if err := o.client.ensureConnected(ctx); err != nil {
+		return nil, "", nil, err
+	}
+
+	// Get the user by username to get their ID
+	users, err := o.client.client.GetUsers(ctx)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	var userID string
+	for _, user := range users {
+		if *user.Username == resource.DisplayName {
+			userID = *user.ID
+			break
+		}
+	}
+
+	if userID == "" {
+		return nil, "", nil, fmt.Errorf("user not found")
+	}
+
+	// Get all groups the user is a member of
+	groups, err := o.client.client.GetUserGroups(ctx, userID)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	for _, group := range groups {
+		groupResource, err := parseIntoGroupResource(group, nil)
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		grant := &v2.Grant{
+			Id: fmt.Sprintf("grant:%s:%s", userID, *group.ID),
+			Entitlement: &v2.Entitlement{
+				Id:          fmt.Sprintf("user:%s:group:%s", userID, *group.ID),
+				DisplayName: fmt.Sprintf("Membership in %s", *group.Name),
+				Description: fmt.Sprintf("Membership in the %s group", *group.Name),
+				GrantableTo: []*v2.ResourceType{userResourceType},
+				Slug:        "membership",
+				Resource:    groupResource,
+			},
+			Principal: resource,
+		}
+
+		grants = append(grants, grant)
+	}
+
+	return grants, "", annos, nil
+}
+
+// newUserBuilder creates a new instance of userBuilder.
+// This is the constructor function for the userBuilder struct.
+func newUserBuilder(client *Connector) *userBuilder {
+	return &userBuilder{
+		resourceType: userResourceType,
+		client:       client,
+	}
+}
+
+// parseIntoUserResource converts a Linode user object into a Baton SDK user resource.
+// Parameters:
+//   - ctx: Context (currently unused)
+//   - user: Pointer to the Linode user object to convert
+//   - parentResourceID: Optional parent resource ID for hierarchy
+//
+// Returns:
+//   - *v2.Resource: The converted Baton resource
+//   - error: Any conversion error that occurred
+func parseIntoUserResource(user *gocloak.User, parentResourceID *v2.ResourceId) (*v2.Resource, error) {
+	var userStatus = v2.UserTrait_Status_STATUS_ENABLED
+
+	username := ""
+	if user.Username != nil {
+		username = *user.Username
+	}
+
+	profile := map[string]interface{}{
+		"username":  username,
+		"email":     safeString(user.Email),
+		"firstName": safeString(user.FirstName),
+		"lastName":  safeString(user.LastName),
+	}
+
+	userTraits := []resource.UserTraitOption{
+		resource.WithUserProfile(profile),
+		resource.WithUserLogin(username),
+		resource.WithStatus(userStatus),
+	}
+
+	ret, err := resource.NewUserResource(
+		username,
+		userResourceType,
+		username,
```

Comment:
We usually use the ID for the 3rd argument here. If you do that, you won't need to list all the users in the `Grants` method (since you can get the ID from `resource.Id.Resource`), which should help performance